contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,621 @@
+            maxAmountInputToken = abi.decode(_paymentParams, (uint256));
+    }
+
+
+    /**
+     * Transfers the shortfall between the amount of tokens required to return flashloan and what was obtained
+     * from swapping the debt tokens from the users address
+     *
+     * @param _token                 Address of the token to transfer from user
+     * @param _amountRequired        Amount required to repay flashloan
+     * @param _amountObtained        Amount obtained from swapping the short token
+     * @param _originalSender        Adress that initiated the token issuance, which is the adresss form which to transfer the tokens
+     */
+    function _transferShortfallFromSender(address _token, uint256 _amountRequired, uint256 _amountObtained, address _originalSender, bytes memory paymentParams) internal returns(uint256 shortfall) {
+        if(_amountObtained < _amountRequired){ 
+            shortfall = _amountRequired.sub(_amountObtained);
+            uint256 maxAmountInputToken =_decodePaymentParamsLongToken(paymentParams);
+            require(shortfall <= maxAmountInputToken, "ExchangeIssuance: INSUFFICIENT INPUT AMOUNT");
+            IERC20(_token).safeTransferFrom(_originalSender, address(this), shortfall);
+        }
+    }
+
+
+
+    /**
+     * Swaps the debt tokens obtained from issuance for the underlying of the collateral
+     *
+     * @param _setToken                   Address of the SetToken to be issued
+     * @param _setAmount                  Amount of SetTokens to issue
+     * @param _longTokenUnderlying        Address of the underlying of the collateral token
+     */
+    function _swapShortForLongTokenUnderlying(address _setToken, uint256 _setAmount, address _longTokenUnderlying) internal returns (uint256) {
+        (, , address shortToken, uint shortAmount) = getLeveragedTokenData(ISetToken(_setToken), _setAmount);
+        (, Exchange exchange, ) = _getMaxTokenForExactToken(shortAmount, shortToken, _longTokenUnderlying);
+        IUniswapV2Router02 router = _getRouter(exchange);
+        _safeApprove(IERC20(shortToken), address(router), shortAmount);
+        return _swapExactTokensForTokens(exchange, shortToken, _longTokenUnderlying, shortAmount);
+    }
+
+    /**
+     * Deposit underlying of collateral to obtain actual collateral token for issuance
+     *
+     * @param _longTokenUnderlying    Address of underlying of the collateral token
+     * @param _depositAmount          Amount to deposit
+     */
+    function _depositLongToken(
+        address _longTokenUnderlying,
+        uint256 _depositAmount
+    ) internal {
+        IERC20(_longTokenUnderlying).approve(address(LENDING_POOL), _depositAmount);
+        LENDING_POOL.deposit(_longTokenUnderlying, _depositAmount, address(this), 0);
+    }
+
+
+    /**
+     * Approves lending pool to transfer owed amount of borrowed tokens after flashloan operation is complete
+     *
+     */
+    function _approveAssetToReturn(
```

Comment:
We can probably do this in `approveSetToken` (and we might want to remove `approveToken` since its probably not important to be able to approve one by one). This prevents us from doing a new approval on each issuance which saves a but of gas (not that it matters much on optimism)